### PR TITLE
Add Half support for softmax and log_softmax on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Add Half support for softmax and log_softmax on CPU.
Note: This introduces a correctness issue with MPS https://github.com/pytorch/pytorch/issues/111416 and https://github.com/pytorch/pytorch/issues/111479.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -10129,6 +10105,34 @@ def test_masked_softmax_TxT_layout(self, device):
         pt_res = self._slow_masked_softmax(input, mask)
         self.assertEqual(pt_res, native_res, exact_dtype=True)
 
+    @onlyCPU
+    @dtypes(torch.bfloat16, torch.half)
+    def test_log_softmax_cpu(self, device, dtype):
```
Comment (jgong5) [inline test/test_nn.py:10110]: why move it to a different place? It makes reviewing harder. :-)

**Review by jgong5 — APPROVED** [review]

File: test/test_nn.py (hunk 1 above)
Comment (CaoE) [reply test/test_nn.py:10246]: Because it can use `@dtypes` in the original place.

**Comment by CaoE** [comment]
@mruberry Could you please review this PR ?

**Comment by CaoE** [comment]
@mikaylagawarecki Could you please review this PR ? Thanks.

File: test/inductor/test_torchinductor_opinfo.py [hunk 2]
```diff
@@ -163,6 +163,9 @@ def format_op(op):
     "linalg.ldl_factor": {f32, f64},  # flaky
     "__rdiv__": {b8, f16, f32, f64, i32, i64},  # flaky
     "nn.functional.cosine_embedding_loss": {b8},  # flaky
+    "_softmax_backward_data": {
+        f16
+    },  # half_to_float is only valid for the CUDA implementatio
```
Comment (mikaylagawarecki) [inline test/inductor/test_torchinductor_opinfo.py:168]: nit:
```suggestion
    },  # half_to_float is only valid for the CUDA implementation
```

**Review by mikaylagawarecki — APPROVED** [review]
Looks ok to me other than a question about `half_to_float`

File: test/inductor/test_torchinductor_opinfo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CaoE) [reply test/inductor/test_torchinductor_opinfo.py:168]: it is useful for preventing data type overflows. https://pytorch.org/docs/stable/generated/torch.nn.functional.softmax.html#torch.nn.functional.softmax 
CPU hasn't support `half_to_float` yet. We may support it in later PRs.
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/SoftMax.cpp#L96-#L102
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/SoftMax.cpp#L456-#L461

[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply test/inductor/test_torchinductor_opinfo.py]: Sounds good, please fix lint

[the remark this answers is not in the document]
Comment (CaoE) [reply test/inductor/test_torchinductor_opinfo.py]: Fixed

**Review by malfet — APPROVED** [review]
Please add a note to PR description that this introduces a correctness issue with MPS
